Teach TextBreakIterator about strict and loose line breaking behaviors

litherum opened the pull request:
#### 99eac467073eccfa5ff272ce4c8212273b2c1040
<pre>
Teach TextBreakIterator about strict and loose line breaking behaviors
<a href="https://bugs.webkit.org/show_bug.cgi?id=257108">https://bugs.webkit.org/show_bug.cgi?id=257108</a>
rdar://109634206

Reviewed by Cameron McCormack.

This is the first part of our line breaker refactoring. We currently have 2 totally separate classes,
both of which can perform line breaking: TextBreakIterator and LazyLineBreakIterator.

TextBreakIterator has multiple backends - it can be backed by either ICU or Core Foundation. It also
supports all kinds of text segmentation: caret positions, grapheme cluster segmentation, etc.

LazyLineBreakIterator only works with ICU, and can only do line breaking. However, it supports loose
and strict line breaking modes, and supports a &quot;prior context.&quot;

It&apos;s kind of a shame that we have 2 classes which both do similar things, so I&apos;m going to try to
unify them into a single class which can do everything. I&apos;d like to improve TextBreakIterator to be
able to do everything that LazyLineBreakIterator can do, and then delete LazyLineBreakIterator,
because I think that&apos;s going to be the most straightforward way of doing it.

This patch teaches TextBreakIterator about the different line breaking behaviors, by turning the
&quot;mode&quot; enum into a variant, and giving the Line struct a behavior enum.

* Source/WTF/wtf/text/TextBreakIterator.cpp:
(WTF::mapModeToBackingIterator):
* Source/WTF/wtf/text/TextBreakIterator.h:
(WTF::TextBreakIterator::LineMode::operator== const):
(WTF::TextBreakIterator::CaretMode::operator== const):
(WTF::TextBreakIterator::DeleteMode::operator== const):
(WTF::TextBreakIterator::CharacterMode::operator== const):
(WTF::LazyLineBreakIterator::get):
(WTF::TextBreakIteratorCache::TextBreakIteratorCache): Deleted.
* Source/WTF/wtf/text/WTFString.h:
(WTF::StringLiterals::operator _str):
* Source/WTF/wtf/text/cocoa/TextBreakIteratorInternalICUCocoa.cpp:
(WTF::mapModeToBackingIterator):
* Source/WTF/wtf/text/icu/TextBreakIteratorICU.h:
(WTF::TextBreakIteratorICU::TextBreakIteratorICU):
(WTF::TextBreakIteratorICU::makeLocaleWithBreakKeyword):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::offsetForPosition):
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/ComposedCharacterClusterTextIterator.h:
(WebCore::ComposedCharacterClusterTextIterator::ComposedCharacterClusterTextIterator):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::previousOffset const):
(WebCore::RenderText::previousOffsetForBackwardDeletion const):
(WebCore::RenderText::nextOffset const):
* Tools/TestWebKitAPI/Tests/WTF/TextBreakIterator.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/264376@main">https://commits.webkit.org/264376@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d817999d9c216739503150f32dd7dafceaf79e4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7451 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7716 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7893 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9087 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/7654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7462 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9653 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7638 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/10541 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7582 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8699 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/6903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9197 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6006 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/6776 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/6349 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/7232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/6894 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10206 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/7035 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7389 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/7597 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6735 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1745 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1760 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10946 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/7800 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7122 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1885 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WTF/wtf/text/TextBreakIterator.h:
```diff
@@ -81,6 +90,9 @@ class TextBreakIterator {
 private:
+    using Backing = std::variant<TextBreakIteratorICU, TextBreakIteratorPlatform>;
+
+    friend Backing mapModeToBackingIterator(StringView, TextBreakIterator::Mode, const AtomString& locale);
```
Nit: Making this a friend just so it can access `TextBreakIterator::Backing` feels a bit strange. A static member function would be more idiomatic.

heycam commented on Source/WTF/wtf/text/TextBreakIterator.h:
```diff
@@ -46,12 +46,21 @@ class TextBreakIteratorCache;
+    struct CharacterMode {
+        bool operator==(const CharacterMode&) const = default;
+    };
+    using Mode = std::variant<LineMode, CaretMode, DeleteMode, CharacterMode>;
```
(Wish C++ had built in sum types...)

heycam commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -566,6 +566,11 @@ inline String operator"" _str(const char* characters, size_t)
     return ASCIILiteral::fromLiteralUnsafe(characters);
 }
 
+inline String operator"" _str(const UChar* characters, size_t length)
+{
+    return String(characters, length);
+}
```
Do we need another type like ASCIILiteral but for UTF-16 strings?

litherum replied:
Possibly/probably, but I think it's outside the scope of this patch.

heycam commented on Source/WTF/wtf/text/icu/TextBreakIteratorICU.h:
```diff
@@ -141,6 +147,54 @@ class TextBreakIteratorICU {
     }
 
 private:
+    // This is temporarily duplicated from LineBreakIteratorPool.
+    // Eventually, LineBreakIteratorPool will be replaced with TextBreakIteratorCache,
+    // and this won't be duplicated any more.
+    static AtomString makeLocaleWithBreakKeyword(const AtomString& locale, LineMode::Behavior behavior)
```
Appreciate that it's temporary, but do we need to duplicate? What about defining `LineMode::Behavior` as an alias for `LineBreakIteratorMode`? That should let us call the existing function. Unless there are other behavior changes in this function (aside from the early return optimization) that I'm overlooking.

heycam commented on Tools/TestWebKitAPI/Tests/WTF/TextBreakIterator.cpp:
```diff
@@ -143,4 +143,26 @@ TEST(WTF, TextBreakIteratorNumCodeUnitsInGraphemeClusters)
+        auto string = u"ぁぁ"_str;
+        auto locale = AtomString("ja"_str);
+        CachedTextBreakIterator strictIterator({ string }, TextBreakIterator::LineMode { TextBreakIterator::LineMode::Behavior::Strict }, locale);
+        EXPECT_EQ(string.length(), strictIterator.following(0));
```
IMO it's slightly better to write it as `EXPECT_EQ(strictIterator.following(0), string.length())`, so that it emphasizes which part of the assertion is the result and which is the expected value.

Also I think it'd be fine, and probably clearer, to hard code in the expected values (i.e. 2 and 1) for the two tests in here.

heycam reviewed (approved):
In the commit message:

> LazyLineBreakIterator only work with ICU, and only can do line breaking.

s/only work/only works/
s/only can do/can only do/

litherum committed:
Teach TextBreakIterator about strict and loose line breaking behaviors
https://bugs.webkit.org/show_bug.cgi?id=257108
rdar://109634206

Reviewed by Cameron McCormack.

This is the first part of our line breaker refactoring. We currently have 2 totally separate classes,
both of which can perform line breaking: TextBreakIterator and LazyLineBreakIterator.

TextBreakIterator has multiple backends - it can be backed by either ICU or Core Foundation. It also
supports all kinds of text segmentation: caret positions, grapheme cluster segmentation, etc.

LazyLineBreakIterator only works with ICU, and can only do line breaking. However, it supports loose
and strict line breaking modes, and supports a "prior context."

It's kind of a shame that we have 2 classes which both do similar things, so I'm going to try to
unify them into a single class which can do everything. I'd like to improve TextBreakIterator to be
able to do everything that LazyLineBreakIterator can do, and then delete LazyLineBreakIterator,
because I think that's going to be the most straightforward way of doing it.

This patch teaches TextBreakIterator about the different line breaking behaviors, by turning the
"mode" enum into a variant, and giving the Line struct a behavior enum.

* Source/WTF/wtf/text/TextBreakIterator.cpp:
(WTF::mapModeToBackingIterator):
* Source/WTF/wtf/text/TextBreakIterator.h:
(WTF::TextBreakIterator::LineMode::operator== const):
(WTF::TextBreakIterator::CaretMode::operator== const):
(WTF::TextBreakIterator::DeleteMode::operator== const):
(WTF::TextBreakIterator::CharacterMode::operator== const):
(WTF::LazyLineBreakIterator::get):
(WTF::TextBreakIteratorCache::TextBreakIteratorCache): Deleted.
* Source/WTF/wtf/text/WTFString.h:
(WTF::StringLiterals::operator _str):
* Source/WTF/wtf/text/cocoa/TextBreakIteratorInternalICUCocoa.cpp:
(WTF::mapModeToBackingIterator):
* Source/WTF/wtf/text/icu/TextBreakIteratorICU.h:
(WTF::TextBreakIteratorICU::TextBreakIteratorICU):
(WTF::TextBreakIteratorICU::makeLocaleWithBreakKeyword):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::offsetForPosition):
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/ComposedCharacterClusterTextIterator.h:
(WebCore::ComposedCharacterClusterTextIterator::ComposedCharacterClusterTextIterator):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::previousOffset const):
(WebCore::RenderText::previousOffsetForBackwardDeletion const):
(WebCore::RenderText::nextOffset const):
* Tools/TestWebKitAPI/Tests/WTF/TextBreakIterator.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/264376@main
